## Preprocessing

In [73]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd
application_df = pd.read_csv("https://static.bc-edx.com/data/dl-1-2/m21/lms/starter/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [74]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
application_df = application_df.drop(['EIN', 'NAME'], axis=1)
application_df.head()

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0              T10       Independent          C1000    ProductDev   
1               T3       Independent          C2000  Preservation   
2               T5  CompanySponsored          C3000    ProductDev   
3               T3  CompanySponsored          C2000  Preservation   
4               T3       Independent          C1000     Heathcare   

   ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  \
0   Association       1              0                      N     5000   
1  Co-operative       1         1-9999                      N   108590   
2   Association       1              0                      N     5000   
3         Trust       1    10000-24999                      N     6692   
4         Trust       1  100000-499999                      N   142590   

   IS_SUCCESSFUL  
0              1  
1              1  
2              0  
3              1  
4              1

In [75]:
# Determine the number of unique values in each column.
application_df.nunique()

APPLICATION_TYPE            17
AFFILIATION                  6
CLASSIFICATION              71
USE_CASE                     5
ORGANIZATION                 4
STATUS                       2
INCOME_AMT                   9
SPECIAL_CONSIDERATIONS       2
ASK_AMT                   8747
IS_SUCCESSFUL                2
dtype: int64

In [76]:
# Look at APPLICATION_TYPE value counts to identify and replace with "Other"
app_counts =  application_df['APPLICATION_TYPE'].value_counts()
app_counts

APPLICATION_TYPE
T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: count, dtype: int64

In [77]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
application_types_to_replace = list(app_counts[app_counts < 500].index)

# Replace in dataframe
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure replacement was successful
application_df['APPLICATION_TYPE'].value_counts()

APPLICATION_TYPE
T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: count, dtype: int64

In [78]:
# Look at CLASSIFICATION value counts to identify and replace with "Other"
application_df['CLASSIFICATION'].value_counts()

CLASSIFICATION
C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: count, Length: 71, dtype: int64

In [79]:
# You may find it helpful to look at CLASSIFICATION value counts >1
application_df['CLASSIFICATION'].value_counts()[application_df['CLASSIFICATION'].value_counts() > 1]

CLASSIFICATION
C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C1800       15
C6000       15
C1250       14
C8200       11
C1238       10
C1278       10
C1235        9
C1237        9
C7210        7
C2400        6
C1720        6
C4100        6
C1257        5
C1600        5
C1260        3
C2710        3
C0           3
C3200        2
C1234        2
C1246        2
C1267        2
C1256        2
Name: count, dtype: int64

In [80]:
# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`
class_counts =  application_df['CLASSIFICATION'].value_counts()
classifications_to_replace = list(class_counts[class_counts < 1000].index)

# Replace in dataframe
for cls in classifications_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls,"Other")

# Check to make sure replacement was successful
application_df['CLASSIFICATION'].value_counts()

CLASSIFICATION
C1000    17326
C2000     6074
C1200     4837
Other     2261
C3000     1918
C2100     1883
Name: count, dtype: int64

In [81]:
# Convert categorical data to numeric with `pd.get_dummies`
application_df = pd.get_dummies(application_df, dtype=float)

In [82]:
# Split our preprocessed data into our features and target arrays
X = application_df.drop(['IS_SUCCESSFUL'], axis=1)
y = application_df['IS_SUCCESSFUL']

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

In [83]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Compile, Train and Evaluate the Model

In [84]:
# Second try
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(tf.keras.layers.Dense(units=80, activation='relu', input_dim=len(X_train.columns)))

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=30, activation='relu'))

# Third hidden layer
# nn.add(tf.keras.layers.Dense(units=30 activation='relu'))
nn.add(tf.keras.layers.Dense(units=30, activation="sigmoid"))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))

# Check the structure of the model
nn.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 80)                3520      
                                                                 
 dense_1 (Dense)             (None, 30)                2430      
                                                                 
 dense_2 (Dense)             (None, 30)                930       
                                                                 
 dense_3 (Dense)             (None, 1)                 31        
                                                                 
Total params: 6911 (27.00 KB)
Trainable params: 6911 (27.00 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [85]:
# Compile the model
nn.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

In [86]:
# Train the model
fit_model = nn.fit(X_train_scaled, y_train, epochs=100)

Epoch 1/100
804/804 [==============================] - 3s 2ms/step - loss: 0.5848 - accuracy: 0.7105
Epoch 2/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5535 - accuracy: 0.7297
Epoch 3/100
804/804 [==============================] - 3s 3ms/step - loss: 0.5512 - accuracy: 0.7300
Epoch 4/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5497 - accuracy: 0.7303
Epoch 5/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5486 - accuracy: 0.7313
Epoch 6/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5471 - accuracy: 0.7325
Epoch 7/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5471 - accuracy: 0.7327
Epoch 8/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5466 - accuracy: 0.7331
Epoch 9/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5460 - accuracy: 0.7315
Epoch 10/100
804/804 [==============================] - 3s 4ms/step - loss: 0.5452 - accura

In [87]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.5622 - accuracy: 0.7279 - 593ms/epoch - 2ms/step
Loss: 0.56218421459198, Accuracy: 0.7279300093650818


### After the second attempt, the model was further optimized, achieving an accuracy of 72.79%. In this iteration, a deep neural network with three hidden layers was defined: the first hidden layer contained 80 units with a ReLU activation function, the second hidden layer had 30 units with a ReLU activation function, and the third hidden layer included 30 units with a sigmoid activation function. The output layer consisted of a single unit with a sigmoid activation function. Despite the adjustments, the model's accuracy improved slightly but remained below the target of 75%. There is still a need for further optimization to reach the desired accuracy.


## Compile, Train and Evaluate the Model##

In [88]:
# Third try
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(tf.keras.layers.Dense(units=110, activation='relu', input_dim=len(X_train.columns)))

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=45, activation='relu'))

# Third hidden layer
nn.add(tf.keras.layers.Dense(units=76, activation="sigmoid"))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))

# Check the structure of the model
nn.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_4 (Dense)             (None, 110)               4840      
                                                                 
 dense_5 (Dense)             (None, 45)                4995      
                                                                 
 dense_6 (Dense)             (None, 76)                3496      
                                                                 
 dense_7 (Dense)             (None, 1)                 77        
                                                                 
Total params: 13408 (52.38 KB)
Trainable params: 13408 (52.38 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [89]:
# Compile the model
nn.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

In [90]:
# Train the model
fit_model = nn.fit(X_train_scaled, y_train, epochs=100)

Epoch 1/100
804/804 [==============================] - 5s 4ms/step - loss: 0.5694 - accuracy: 0.7215
Epoch 2/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5550 - accuracy: 0.7276
Epoch 3/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5523 - accuracy: 0.7315
Epoch 4/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5513 - accuracy: 0.7283
Epoch 5/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5496 - accuracy: 0.7306
Epoch 6/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5492 - accuracy: 0.7321
Epoch 7/100
804/804 [==============================] - 3s 4ms/step - loss: 0.5476 - accuracy: 0.7333
Epoch 8/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5474 - accuracy: 0.7324
Epoch 9/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5464 - accuracy: 0.7320
Epoch 10/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5459 - accura

In [91]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - loss: 0.5761 - accuracy: 0.7285 - 469ms/epoch - 2ms/step
Loss: 0.5761218667030334, Accuracy: 0.7285131216049194


In [93]:
# Export our model to HDF5 file
nn.save('AlphabetSoupCharity_Optimization.h5.')

### After the third attempt, the model was optimized and achieved an accuracy of 73%. This represents an improvement but still falls short of the target accuracy of 75%. In this attempt, a deep neural network with three hidden layers was defined: the first hidden layer had 110 units with a ReLU activation function, the second hidden layer had 45 units with a ReLU activation function, and the third hidden layer had 76 units with a sigmoid activation function. The output layer consisted of a single unit with a sigmoid activation function. Despite these optimizations, the model's performance indicates that further enhancements are necessary.

### **Random Forest**

In [38]:
# Create and fit the Random Forest model
rf_model = RandomForestClassifier(n_estimators=150, random_state=75)
rf_model.fit(X_train_scaled, y_train)

# Evaluate feature importance
importances = rf_model.feature_importances_
feature_importance = pd.Series(importances, index=X_RF.columns).sort_values(ascending=False)
print("Feature Importances:")
print(feature_importance)

# Make predictions on the test set
y_pred = rf_model.predict(X_test_scaled)

Feature Importances:
ASK_AMT                         0.408239
AFFILIATION_CompanySponsored    0.133040
AFFILIATION_Independent         0.119157
ORGANIZATION_Association        0.029827
APPLICATION_TYPE_T5             0.027242
APPLICATION_TYPE_T10            0.023767
ORGANIZATION_Trust              0.021335
CLASSIFICATION_C2100            0.020155
CLASSIFICATION_Other            0.018729
APPLICATION_TYPE_T19            0.017630
CLASSIFICATION_C2000            0.017594
APPLICATION_TYPE_T4             0.017013
APPLICATION_TYPE_T3             0.014418
CLASSIFICATION_C1000            0.013140
APPLICATION_TYPE_T6             0.010696
USE_CASE_ProductDev             0.010519
INCOME_AMT_0                    0.009949
CLASSIFICATION_C1200            0.009521
USE_CASE_Preservation           0.009363
CLASSIFICATION_C3000            0.008243
INCOME_AMT_25000-99999          0.006953
INCOME_AMT_1M-5M                0.006538
INCOME_AMT_1-9999               0.006349
INCOME_AMT_100000-499999        0.00

In [41]:
# Evaluate the model
accuracy = accuracy_score(y_test, y_pred)
print(f"Random forest model accuracy: {accuracy:.3f}")

Random forest model accuracy: 0.721


In [42]:
# Calculate the confusion matrix
cm = confusion_matrix(y_test, y_pred)
cm_df = pd.DataFrame(cm, index=["Actual 0", "Actual 1"], columns=["Predicted 0", "Predicted 1"])

# Display the confusion matrix
print("Confusion Matrix:")
print(cm_df)

Confusion Matrix:
          Predicted 0  Predicted 1
Actual 0         2236         1023
Actual 1          892         2709


In [43]:
# Display the classification report
print("Classification Report:")
print(classification_report(y_test, y_pred))

Classification Report:
              precision    recall  f1-score   support

           0       0.71      0.69      0.70      3259
           1       0.73      0.75      0.74      3601

    accuracy                           0.72      6860
   macro avg       0.72      0.72      0.72      6860
weighted avg       0.72      0.72      0.72      6860



### The Random Forest model was evaluated with the following results: The model achieved an accuracy of 72.1%, reflecting its effectiveness in classifying application success. The confusion matrix indicated that the model correctly predicted 2,236 instances of class 0 and 2,709 instances of class 1, while misclassifying 1,023 instances of class 0 as class 1 and 892 instances of class 1 as class 0. The classification report detailed the model's performance with a precision of 0.71 and recall of 0.69 for class 0, and a precision of 0.73 and recall of 0.75 for class 1. The F1-scores were 0.70 for class 0 and 0.74 for class 1. With a macro and weighted average of approximately 0.72 across precision, recall, and F1-score, the results show a balanced performance. Despite these results, further optimization may be necessary to improve the model’s accuracy and achieve better predictive performance.

# LogisticRegression

In [65]:
# Create and train the Logistic Regression model
log_reg_model = LogisticRegression(max_iter=1000, random_state=75)
log_reg_model.fit(X_train_scaled, y_train)

LogisticRegression(max_iter=1000, random_state=75)

In [66]:
# Predict the target values for the test set
y_pred = log_reg_model.predict(X_test_scaled)

# Evaluate the model's accuracy
accuracy = accuracy_score(y_test, y_pred)
print(f"Logistic Regression model accuracy: {accuracy:.3f}")

# Calculate the confusion matrix
cm = confusion_matrix(y_test, y_pred)
cm_df = pd.DataFrame(cm, index=["Actual 0", "Actual 1"], columns=["Predicted 0", "Predicted 1"])

# Display the confusion matrix
print("Confusion Matrix:")
print(cm_df)

# Display the classification report
print("Classification Report:")
print(classification_report(y_test, y_pred))

Logistic Regression model accuracy: 0.722
Confusion Matrix:
          Predicted 0  Predicted 1
Actual 0         2711         1282
Actual 1         1105         3477
Classification Report:
              precision    recall  f1-score   support

           0       0.71      0.68      0.69      3993
           1       0.73      0.76      0.74      4582

    accuracy                           0.72      8575
   macro avg       0.72      0.72      0.72      8575
weighted avg       0.72      0.72      0.72      8575



### The Logistic Regression model was developed and trained with a maximum of 1000 iterations and a fixed random seed of 75. On evaluation with the test data, the model achieved an accuracy of 72.2%. The confusion matrix revealed that the model correctly predicted 2,711 instances of class 0 and 3,477 instances of class 1, while misclassifying 1,282 instances of class 0 as class 1 and 1,105 instances of class 1 as class 0. The classification report further detailed the model's performance, showing a precision of 0.71 and recall of 0.68 for class 0, and a precision of 0.73 and recall of 0.76 for class 1. The F1-scores were 0.69 for class 0 and 0.74 for class 1. The macro and weighted averages for precision, recall, and F1-score were all around 0.72, indicating balanced performance across classes. Although the model demonstrates a reasonable level of accuracy, further optimization may be necessary to achieve higher performance benchmarks.